### PR TITLE
perf: Mobject's Sequence methods no longer copy self.submobjects on every call

### LINKED ISSUE #4914: Mobject: __iter__, __len__, and __getitem__ make a copy of self.submobjects on every call
## Problem description
All three of `Mobject.__iter__`, `Mobject.__len__`, and `Mobject.__getitem__` operate on the output of `self.split()`, which is implemented like so:
```py
def split(self) -> list[Mobject]:
    result: list[Mobject] = [self] if len(self.points) > 0 else []
    return result + self.submobjects
```
As a result, a new list created from `self.submobjects` (and possibly `self`) is built _every time_ one of these methods is called, leading to an unnecessary use of both time and memory. The problem is especially egregious in the case of `__len__` and `__getitem__`, which reasonably should have constant-time access but instead have O(N) time (and space?) complexity for N submobjects.

## Benchmark
As a super quick test, I measured the running time of constructing each of the three lists in the following:
```py
# PseudoPython; I did this more intelligently IRL
for n in np.linspace(10, 200_000, 10):
    mob = VGroup(Circle() for _ in range(n))

    # measure each of these
    a = [*mob]
    b = [*mob.submobjects]
    c = [mob[i] for i in range(len(mob.submobjects))]
```
The results look like this:
Image
I haven't measured memory usage.

## Other comments
Curiously, this only seems to be an issue for `Mobject`; in `OpenGLMobject`, `split` is implemented as follows:
```py
def split(self) -> Sequence[OpenGLMobject]:
    return self.submobjects
```
It's not clear to me whether this different in the behavior of `split` is intended or accidental. That's maybe one for a separate issue.

### PR DESCRIPTION
**Pull request by nikolajmunk**
## Overview: What does this pull request change?
Adresses performance issues caused by always calling `self.split()` (which creates a copy of the submobject list) in the following methods:
- Turn `__iter__` into an actual iterator instead of a copied list
- Make `__len__` constant-time by just summing list lengths
- Simulate `__getitem__` indexing into `self.split()` (both for int and slice values) without actually creating it

Closes #4914.

## Motivation and Explanation: Why and how do your changes improve the library?

Dramatically improves performance of the above methods when `self.submobjects` is large. `__getitem__` is particularly improved; some casual benchmarking shows that single-index indexing is many, many times faster than the current implementation, and slicing is roughly twice as fast (though much of the performance improvement is eaten by the overhead of constructing the `GroupClass`.

## Further Information and Comments

It seems to me that there aren't any unit tests for this kind of peculiar behavior of `Mobject` (behave as `self.submobjects` if the mobject has no points and `[self] + self.submobjects` if it does). I'm happy to write some if people feel it necessary. This implementation _is_ more complicated than the current one, so it might not be a bad idea.

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by nikolajmunk** [comment]
I'm not married to the `return cast(Mobject | Group, self.submobjects[value])` thing, but it seemed a nice enough way to satisfy the type checker.

I wonder if it makes sense to completely drop the `Mobject | Group` union from the return type since `Group` is a subclass of `Mobject` anyway? Then I believe we should be able to return everything without an explicit cast.

**Comment by behackl** [comment]
Ohh, this looks neat too! I'd be happy with just returning this as `Mobject`.

**Comment by pjfo** [comment]
I'm running my large scene bench against this, since it contains many objects.

I threw an LLM at it, got it to validate with a repro script that tested the old-split() based semantics on 0.20.1 vs the new way provided by the commit for this PR, and it found 4 potential issues (the repro script is detailed at the end). 

What follows is the LLM description of the issues:
1. Out-of-range integer indices that are multiples of len(self) return self instead of raising IndexError

value % len(self) == 0 is true for every multiple of the virtual length, but only 0 and -len(self) are valid indices resolving to self:

m = Mobject(); m.points = np.zeros((3, 3)); m.add(a, b, c)   # split() == [m, a, b, c]
m.split()[4]   # IndexError
m[4]           # returns m
m[8], m[-8], m[100]  # likewise return m

For a pointful mobject with no submobjects (len(m) == 1), every integer index now returns self.

2. Empty slices return non-empty Groups

When the normalized range is empty, the translated bounds wrap around to negative indices (and the stop = None case re-includes elements):

m.split()[1:0]      # []
m[1:0]              # Group(a, b)
m[2:-100]           # Group(b)
m[0:0:-1]           # Group(c, b, a)
m[-4:0:-1]          # Group(c, b, a)

3. The % fast path breaks the __index__ protocol

Lists accept any object implementing __index__ and reject floats; % does the opposite:

m[0.0]          # returns m — old behavior: TypeError
class IndexOne:
    def __index__(self): return 1
m[IndexOne()]   # TypeError — old behavior: first submobject

4. __iter__ no longer iterates a snapshot

iter(self.split()) iterated a fresh list, so mutating submobjects mid-iteration was safe. The chain iterates the live list:

for sub in m:        # m has 4 submobjects
    m.remove(sub)    # old: removes all 4; new: skips alternate elements, 2 remain

If live iteration is intended it's a behavioral change worth calling out explicitly rather than a pure perf refactor; if not, chaining over a copy of submobjects keeps the snapshot semantics while still avoiding the split() call.

Suggested fixes

For 1 + 3 together, the integer path can delegate normalization, bounds-checking, and __index__ handling to range:

index = range(len(self))[value]   # raises IndexError/TypeError exactly like list indexing
if self.has_no_points():
    return self.submobjects[index]
return self if index == 0 else self.submobjects[index - 1]

For 2, early-return GroupClass() when len(r) == 0 — the bound-translation arithmetic is only valid for non-empty ranges (and as far as I can tell it is correct for all of them, including [::-1], clamped bounds, and step > 1).

A related design note: these double underscore methods previously routed through self.split(), so a subclass overriding split() also controlled indexing/iteration/length; that's no longer the case. Probably fine, but worth a conscious decision.

Repro script:
[pr4915_repro.py](https://github.com/user-attachments/files/30747415/pr4915_repro.py)

0.20.1 output:
[output_v0.20.1_release.txt](https://github.com/user-attachments/files/30747373/output_v0.20.1_release.txt)

commit 16a68aa output:
[output_pr4915_16a68aa.txt](https://github.com/user-attachments/files/30747403/output_pr4915_16a68aa.txt)


(Note: it is my intention to run the repro script against the current main head later once the large scene bench completes)

**Comment by nikolajmunk** [comment]
Wow @pjfo, that's a mouthful!

Good catch on issues 1 and 2, I thought I had handled both cases but I think it got lost in some other change.

3 seems like a trivial fix, it makes sense that the type of `value` should be something like `SupportsIndex | slice[SupportsIndex | None]` (at least for non-dictlike implementations). What do people think?

As for 4: personally I think `__iter__` is fine as it is; the default behavior of `list` is to permit mutation of itself while iterating (one of my favorite tricks is to implement BFS by appending to a list while iterating over it), so we might as well do the same. Happy to hear arguments for other choices.

edit: OK, I think I've got fixes for 1-3. Let me test them properly (the repro script above is happy!) and push the change.

**Comment by nikolajmunk** [comment]
Points 1 and 2 maybe also emphasize that we need a test for these methods, since the current code managed to pass all tests even with the bugs mentioned here.

File: manim/mobject/mobject.py [hunk 1]
```diff
@@ -2490,18 +2490,35 @@ def align_to(
...
+
+            if self.has_no_points():
+                return GroupClass(*self.submobjects[value])
+
+            r = range(*value.indices(len(self)))
+            if 0 not in r:
+                # If self is not included in the slice, we can gain a small speed boost
+                # by indexing directly into the submobjects list.
+                stop = r.stop - 1 if (r.step > 0 or r.stop > 0) else None
+                return GroupClass(*self.submobjects[r.start - 1 : stop : r.step])
```
Comment (GniLudio) [inline manim/mobject/mobject.py:2506]: I think `0 not in r` has a cost of `O(n)`. So it it really a speed boost?

Comment (nikolajmunk) [reply manim/mobject/mobject.py:2506]: AFAICT, `x in range(...)` is O(1) when `x` is an integer. I had the same thought originally so I Googled around a bit first. [Here's one source](https://stackoverflow.com/questions/30081275/why-is-1000000000000000-in-range1000000000000001-so-fast-in-python-3/30081470#30081470).

Comment (GniLudio) [reply manim/mobject/mobject.py:2506]: That makes sense.

File: manim/mobject/mobject.py [hunk 2]
```diff
@@ -2490,18 +2490,35 @@ def align_to(
     # Family matters
 
     def __getitem__(self, value: Any) -> Mobject | Group:
```
Comment (GniLudio) [inline manim/mobject/mobject.py]: This would have the same behavior as list slicing:
```python
    def __getitem__(self, value: Any) -> Mobject | Group:
        if isinstance(value, slice):
            GroupClass = self.get_group_class()
            if self.has_no_points():
                return GroupClass(*self.submobjects[value])
            r = range(*value.indices(len(self)))
            return GroupClass(*[self if i == 0 else self.submobjects[i - 1] for i in r])
        if self.has_no_points():
            return cast(Mobject | Group, self.submobjects[value])
        if value == 0 or value == -len(self):
            return self
        return cast(
            Mobject | Group, self.submobjects[value - 1 if value > 0 else value]
        )
```

<details>
<summary>Equality Test</summary>

```py
from manim import *

mobjects: list[VMobject] = [
    Square(),  # or VMObject()
    Circle(),
    Square(),
    Rectangle(),
    Dot(),
    Square(),
    Rectangle(),
]
mob = mobjects[0]
for child in mobjects[1:]:
    mob.add(child)

if mobjects[0].has_no_points():
    mobjects = mobjects[1:]

assert mobjects == list(mob)
assert mobjects[0] == mob[0]
assert mobjects[-1] == mob[-1]
assert mobjects[0 : len(mobjects)] == mob[0 : len(mob)].submobjects
assert mobjects[:] == mob[:].submobjects
assert mobjects[0:] == mob[0:].submobjects
assert mobjects[1:] == mob[1:].submobjects
assert mobjects[:0] == mob[:0].submobjects
assert mobjects[:-1] == mob[:-1].submobjects
assert mobjects[1:5:2] == mob[1:5:2].submobjects
assert mobjects[-1:-5:-2] == mob[-1:-5:-2].submobjects
assert mobjects[-5:-1:-3] == mob[-5:-1:-3].submobjects
# mobjects[100] == mob[100] # IndexError
# mobjects[-100] == mob[-100] # IndexError
```
</details>

There is probably some intelligent way to use slicing for the submobjects, but I haven't found a way that doesn't modify the behavior.

Comment (nikolajmunk) [reply manim/mobject/mobject.py]: Looks good to me. I've got a very similar change written up that I just want to make sure I test properly before I push. I'm happy to discuss details then!

I would personally be a little sad to lose the `if 0 not in r` shortcuts as it does seem like it provides an actual performance benefit at the cost of some elegance, but if we eventually lose it I won't lie awake at night.

Comment (GniLudio) [reply manim/mobject/mobject.py]: Feel free to share an implementation that has that performance optimization while passing the equality test...

Comment (GniLudio) [reply manim/mobject/mobject.py]: An LLM gave me this, which seems to be passing the equality test:
```py
    def __getitem__(self, value: Any) -> Mobject | Group:
        if isinstance(value, slice):
            GroupClass = self.get_group_class()
            if self.has_no_points():
                return GroupClass(*self.submobjects[value])
            r = range(*value.indices(len(self)))
            if not r:
                return GroupClass()
            start = r.start
            stop = r.stop
            step = r.step
            a = max(0, start - 1)
            b = stop - 1 if stop > 0 else None
            c = step
            if step < 0 and stop < 0:
                b = stop
            if 0 in r:
                return GroupClass(self, *self.submobjects[a:b:c])
            return GroupClass(*self.submobjects[a:b:c])
        if self.has_no_points():
            return cast(Mobject | Group, self.submobjects[value])
        if value == 0 or value == -len(self):
            return self
        return cast(
            Mobject | Group, self.submobjects[value - 1 if value > 0 else value]
        )
```

Comment (nikolajmunk) [reply manim/mobject/mobject.py]: Quick counterexample: this seems to break in a few different ways for `assert mob[::-1].submobjects == mobjects[::-1]`.

Comment (GniLudio) [reply manim/mobject/mobject.py]: <details>
<summary><s>What about this?</s> Nevermind.</summary>

```py
    def __getitem__(self, value: Any) -> Mobject | Group:
        if isinstance(value, slice):
            GroupClass = self.get_group_class()
            if self.has_no_points():
                return GroupClass(*self.submobjects[value])
            r = range(*value.indices(len(self)))
            if 0 not in r:
                r = slice(
                    r.start - 1 if r.start > 0 else r.start,
                    r.stop - 1 if r.stop > 0 else r.stop,
                    r.step,
                )
                return GroupClass(*self.submobjects[r])
            return GroupClass(*[self if i == 0 else self.submobjects[i - 1] for i in r])

        if self.has_no_points():
            return cast(Mobject | Group, self.submobjects[value])
        if value == 0 or value == -len(self):
            return self
        return cast(
            Mobject | Group, self.submobjects[value - 1 if value > 0 else value]
        )
```
</details>

Comment (nikolajmunk) [reply manim/mobject/mobject.py]: Gonna resolve this - I just pushed a very similar piece of code :)

**Comment by GniLudio** [comment]
Maybe add some test?
```py
from manim import *

mobjects: list[VMobject] = [
    Square(),
    Circle(),
    Square(),
    Rectangle(),
    Dot(),
    Square(),
    Rectangle(),
]
mob = mobjects[0]
for child in mobjects[1:]:
    mob.add(child)

if mobjects[0].has_no_points():
    mobjects = mobjects[1:]


for a in range(-10, 10, 1):
    if a >= -len(mobjects) and a <= (len(mobjects) - 1):
        assert mobjects[a] == mob[a]
    assert mobjects[a:] == mob[a:].submobjects
    assert mobjects[:a] == mob[:a].submobjects
    for b in range(-10, 10, 1):
        assert mobjects[a:b:] == mob[a:b].submobjects
        if b != 0:
            assert mobjects[a::b] == mob[a::b].submobjects
            assert mobjects[:a:b] == mob[:a:b].submobjects
        for c in range(-10, 10, 1):
            if c != 0:
                assert mobjects[a:b:c] == mob[a:b:c].submobjects
```

**Comment by nikolajmunk** [comment]
All right, this implementation seems to pass my own local tests. `self.__getitem__` _should_ now behave indistinguishably to `self.split().__getitem__`.

I still think this needs proper tests. Are there any stupid edge cases I'm likely to be ignoring and which @GniLudio hasn't covered above?

**Comment by behackl** [comment]
Test(s) please, yes! As far as my clanker and I checked you've covered stuff pretty well here already.

Please just make sure to adapt `VMobject.__iter__` and implement it as

```py
def __iter__(self) -> Iterator[VMobject]:
    return cast(Iterator[VMobject], super().__iter__())
```

or so. Great stuff here!

**Comment by nikolajmunk** [comment]
@behackl, good point on `VMobject.__iter__`. I noticed two other places which make calls directly to `split`.

`SampleSpace`:
```py
def __getitem__(self, index: int) -> VMobject:
    if hasattr(self, "horizontal_parts"):
        return self.horizontal_parts[index]
    elif hasattr(self, "vertical_parts"):
        return self.vertical_parts[index]
    return self.split()[index]
````
and `MathTex`:
```py
def index_of_part(self, part: VMobject) -> int:
    split_self = self.split()
    if part not in split_self:
        raise ValueError("Trying to get index of part not in MathTex")
    return split_self.index(part)
```
Should I go ahead and also adjust those while I'm at it?

**Comment by behackl** [comment]
I think that would make sense. I have no recollection of what SampleSpace is whatsoever, so no idea what the correct behavior is there. My clanker says `return cast(VMobject, super().__getitem__(index))` works there 🤷 

There also is a third relevant call in `Matrix.get_det_text`, there is a `parens.split()` that can be changed to `l_paren, r_paren = parens`  i think.

**Comment by nikolajmunk** [comment]
How are we feeling about these tests? I think they belong in `tests_mobject.py`.
```py
def build_mobject(n=10, has_points=False):
    mob = Mobject()
    mob.add(*[Mobject() for _ in range(n)])
    if has_points:
        mob.points = np.zeros((3, 3))
    return mob, mob.split()

@pytest.mark.parametrize("has_points", [True, False])
def test_getitem_single_index(has_points, n=10):
    """Test that indexing a Mobject behaves the same as indexing Mobject.split()."""
    mob, mob_list = build_mobject(n, has_points=has_points)
    length = (n + 1) if has_points else n
    for i in range(-length, length):
        print(i)
        assert mob[i] == mob_list[i]

    # Check that indexing outside valid range raises IndexError
    with pytest.raises(IndexError):
        mob[length]
    with pytest.raises(IndexError):
        mob[-length - 1]

@pytest.mark.parametrize("has_points", [True, False])
def test_getitem_single_index_no_submobjects(has_points):
    """Test that indexing a Mobject with no submobjects behaves correctly."""
    mob, mob_list = build_mobject(0, has_points=has_points)
    if has_points:
        assert mob[0] == mob
    else:
        with pytest.raises(IndexError):
            mob[0]
    with pytest.raises(IndexError):
        mob[1]

@pytest.mark.parametrize("has_points", [True, False])
def test_getitem_slice(has_points, n=10):
    """Test that slicing a Mobject behaves the same as slicing Mobject.split()."""
    mob, mob_list = build_mobject(n, has_points=has_points)
    length = (n + 1) if has_points else n
    for start, stop, step in product(range(-length, length), repeat=3):
        assert mob[start:stop].submobjects == mob_list[start:stop]
        if step == 0:
            with pytest.raises(ValueError):
                mob[start:stop:step]
        else:
            assert mob[start:stop:step].submobjects == mob_list[start:stop:step]

    # Check that slicing with too high stop value works
    assert mob[: length * 2].submobjects == mob_list[: length * 2]

    # Check that slicing with too low start value works
    assert mob[-(length * 2):].submobjects == mob_list[-(length * 2):]

@pytest.mark.parametrize("has_points", [True, False])
def test_getitem_slice_no_submobjects(has_points):
    """Test that slicing a Mobject with no submobjects behaves correctly."""
    mob, _ = build_mobject(0, has_points=has_points)
    expected = [] if not has_points else [mob]
    if has_points:
        assert mob[:].submobjects == expected
        assert mob[::].submobjects == expected
        assert mob[:1].submobjects == expected
        assert mob[::2].submobjects == expected
    else:
        assert mob[:].submobjects == expected
        assert mob[::].submobjects == expected
        assert mob[:1].submobjects == expected
        assert mob[::2].submobjects == expected

@pytest.mark.parametrize("has_points", [True, False])
@pytest.mark.parametrize("n", [0, 10])
def test_iter_mobject_unpacking(has_points, n):
    """Test that unpacking a Mobject behaves correctly."""
    mob, mob_list = build_mobject(n, has_points=has_points)
    assert list(mob) == mob_list
    assert [*mob] == mob_list

@pytest.mark.parametrize("has_points", [True, False])
@pytest.mark.parametrize("n", [0, 10])
def test_iter_mobject_iteration(has_points, n):
    """Test that iterating over a Mobject behaves the same as iterating over Mobject.split()."""
    mob, mob_list = build_mobject(n, has_points=has_points)
    mob_iterator = iter(mob)
    mob_list_iterator = iter(mob_list)
    for a,b in zip(mob_iterator, mob_list_iterator):
        assert a == b

@pytest.mark.parametrize("has_points", [True, False])
@pytest.mark.parametrize("n", [0, 10])
def test_len_mobject(has_points, n):
    """Test that len(mob) behaves the same as len(mob.split())."""
    mob, mob_list = build_mobject(n, has_points=has_points)
    assert len(mob) == len(mob_list)
```

**Comment by pjfo** [comment]
@nikolajmunk I asked my LLM (out of expediency because I have been busier than usual today), It gave me some more comments:

These tests look good — they cover the shapes and exception paths that mattered most (both has_points variants, the n=0 degenerate cases, out-of-range IndexError, zero-step ValueError, clamped bounds). Two gaps left, plus a couple of nits:

1. Non-int index types. Nothing pins the index-type contract, which is where one of the original bugs lived (m[0.0] returned self). List semantics: everything implementing __index__ works, floats raise:

```python
@pytest.mark.parametrize("has_points", [True, False])
def test_getitem_index_types(has_points, n=10):
    """Indexing accepts anything implementing __index__ and rejects floats, like list."""
    mob, mob_list = build_mobject(n, has_points=has_points)

    class Ix:
        def __index__(self):
            return 1

    assert mob[True] == mob_list[1]
    assert mob[np.int64(2)] == mob_list[2]  # e.g. mob[np.argmin(...)]
    assert mob[Ix()] == mob_list[1]
    with pytest.raises(TypeError):
        mob[1.5]
```
2. Mutation during iteration. Since live iteration is the intentional choice, a test should pin it so it can't regress silently in either direction — and the behavior change deserves a line in the __iter__ docstring/changelog, since code like for sub in m: m.remove(sub) behaved differently before this PR:

```python
def test_iter_is_live():
    """Iteration reflects concurrent submobject mutation (changed in this PR: it
    previously iterated a split() snapshot)."""
    mob, _ = build_mobject(4, has_points=False)
    removed = []
    for sub in mob:
        mob.remove(sub)
        removed.append(sub)
    assert len(removed) == 2  # live-list iteration skips alternate elements

```
Related: VMobject.__iter__ is currently still iter(self.split()) — a snapshot. So switching it to super().__iter__() as suggested above doesn't just fix the element type; it also flips VGroup iteration from snapshot to live, which is where mutate-while-iterating is most likely in user code. Fine either way, but that makes it two classes' worth of behaviour change to document, and the test above should then cover both.

Nits: there's a stray print(i) in test_getitem_single_index, and in test_getitem_slice_no_submobjects the if has_points: and else: branches are identical — the assertions can be flattened since expected already differs.

It ran these against the three different commits:
| test               | 09cfc59a fixed head | 16a68aaf buggy head | b3560c59 merge-base |
| :--- | :--: | :--: | :--: |
| author's 6 tests (21 cases)     |        pass         |       4 fail        |        pass        |
| test_getitem_index_types (draft) |        pass        |    fails [True]     |        pass         |
| test_iter_is_live (draft)        |        pass         |        pass        |        fails        |
| total                            |     23/23 pass      | 4 failed, 19 passed | 1 failed, 22 passed |

File: manim/mobject/graphing/probability.py [hunk 3]
```diff
@@ -207,12 +207,12 @@ def add_braces_and_labels(self) -> None:
...
                     self.add(getattr(parts, subattr))
 
-    def __getitem__(self, index: int) -> VMobject:
+    def __getitem__(self, index: Any) -> VMobject:
         if hasattr(self, "horizontal_parts"):
             return self.horizontal_parts[index]
         elif hasattr(self, "vertical_parts"):
             return self.vertical_parts[index]
-        return self.split()[index]
+        return cast(VMobject, super().__getitem__(index))
```
Comment (nikolajmunk) [inline manim/mobject/graphing/probability.py:215]: I don't _love_ the amount of `cast` I'm doing in these files just to communicate that a `VMobject` method is indeed returning a `VMobject` even if `__getitem__` returns a `Mobject`, etc. Is there some kind of meta-type we can use? Something that evaluates to "`VMobject` if I'm a subclass of `VMobject` and `Mobject` if I'm not"?

**Comment by nikolajmunk** [comment]
I've added the tests mentioned above by myself and @pjfo, with the addition of a test for the type of the mobject group created from slicing. I'm happy to hear suggestions for further tests.

I deliberately did not add a test for the "mutation during iteration" case suggested by @pjfo's AI. I'm not convinced that the original behavior was a conscious choice, so I'm not comfortable cementing the definition of `__iter__`'s intended behavior either way by writing a test for it.

**Comment by GniLudio** [comment]
Can you do a before-after time comparison, @nikolajmunk ?
Like the one in the issue description?

File: manim/mobject/graphing/probability.py (hunk 3 above)
Comment (GniLudio) [reply manim/mobject/graphing/probability.py:215]: One idea would be to declare `Mobject` as a generic class:
```py
class Mobject[T = Mobject]:
    submobjects: list[T]
class VMobject(Mobject[VMobject]):
    ...
```
But that would require changing a lot of things.

**Comment by nikolajmunk** [comment]
Happily!

It actually appears that I was being inadvertently charitable to `__getitem__` in my original benchmark; I hadn't realized that `VGroup` implements its own `__getitem__` which is much more reasonable; I would .

So I ran the script again, this time with
```py
mob = Mobject()
mob.submobjects = [Mobject() for _ in range(n)]
```
and got much, much more dramatic results. Here is a plot of the timings for the _before_ case (i.e. `split()`-based methods):
benchmark_before
And here is one for the branch in its current state:
benchmark_after

I would be very grateful if someone would independently try to time this, I almost don't believe what I'm seeing.

edit: come to think of it, this result probably tracks with the O(N) time complexity of `split()`. Here's the per-call running times just for `__getitem__`. The "before" curve looks linear-ish and the "after" curve looks constant-ish.
benchmark_relative

File: manim/mobject/mobject.py [hunk 4]
```diff
@@ -2489,19 +2489,34 @@ def align_to(
...
+            return GroupClass(*[self.submobjects[i - 1] if i != 0 else self for i in r])
+
+        index: int = range(len(self))[value]  # Normalize the index
+        if self.has_no_points():
+            return self.submobjects[index]
+        return self if index == 0 else self.submobjects[index - 1]
 
     def __iter__(self) -> Iterator[Mobject]:
-        return iter(self.split())
+        return it.chain([self] if self.has_points() else [], self.submobjects)
```
Comment (GniLudio) [inline manim/mobject/mobject.py:2516]: Maybe use this instead?
```py
    def __iter__(self) -> Iterator[Mobject]:
        if self.has_points():
            yield self
        yield from self.submobjects
```

Comment (nikolajmunk) [reply manim/mobject/mobject.py:2516]: I mean that certainly looks very nice, simple and professional! I like it.

Weirdly, a very unscientific comparison suggests that `chain` is actually a teeny bit faster than the generator function. Here's the median running times from 100 trials:
chain_yield_compare
I assume there's C function shenanigans that provide a small speed improvement. But maybe that negligible difference is fine if the generator version is easier to understand? No strong opinion either way.

Comment (GniLudio) [reply manim/mobject/mobject.py:2516]: Both are fine for me.

Such a small speed difference especially doesn't matter for `__iter__`, since that isn't a method that is called constantly. (in constrast to `__len__` and `__getitem__`)

**Comment by GniLudio** [comment]
Here are my benchmark results.
getitem
iter
len

<details>
<summary>Benchmarking Code</summary>

```py
import random
import timeit
from collections.abc import Callable
from typing import Any

import pandas as pd
from manim import *
from matplotlib import pyplot as plt

SUBMOBJECT_COUNTS = [1, 5, 10, 50, 100, 500, 1_000, 5_000, 10_000, 50_000, 100_000]
TIMEIT_NUMBER = 1_000
TESTS: list[tuple[str, Callable[[Mobject, int], Any]]] = [
    ("__len__", lambda mob, _: mob.__len__()),
    ("__getitem__", lambda mob, n: mob.__getitem__(random.randint(0, n - 1))),
    ("__iter__", lambda mob, _: mob.__iter__()),
]


def dump_timings() -> None:
    def create_mobject(n: int) -> Mobject:
        mob = Mobject()
        mob.submobjects = [Mobject() for _ in range(n)]
        return mob

    def run_test(f: Callable[[Mobject, int], Any]) -> list[float]:
        results: list[float] = []
        for n in SUBMOBJECT_COUNTS:
            mob = create_mobject(n)
            times = timeit.timeit(stmt=lambda mob=mob, n=n: f(mob, n), number=TIMEIT_NUMBER)
            results.append(times / TIMEIT_NUMBER)

        return results

    print([run_test(test_function) for _, test_function in TESTS])


def plot_timings() -> None:
    main_results = [
        [1.373999984934926e-07, 1.4100002590566874e-07, 1.4709995593875646e-07, 1.9539997447282075e-07, 2.593999961391091e-07, 7.38999922759831e-07, 1.782699953764677e-06, 9.266800014302135e-06, 2.3012800025753675e-05, 0.00024650660005863756, 0.0005839552999241278],
        [4.2749999556690456e-07, 3.9879989344626665e-07, 4.07499959692359e-07, 4.43299999460578e-07, 5.475999787449837e-07, 1.124999951571226e-06, 2.2823000326752664e-06, 9.84229997266084e-06, 2.0857300027273595e-05, 0.00026491879997774957, 0.0006236098000081256],
        [1.5219999477267264e-07, 1.550000160932541e-07, 1.6149994917213915e-07, 2.006000140681863e-07, 3.141999477520585e-07, 7.71900056861341e-07, 1.6865000361576676e-06, 9.338099975138903e-06, 2.3270800011232495e-05, 0.00023934480000752955, 0.000570793699938804],
    ]
    fixed_results = [
        [9.290000889450312e-08, 1.0349997319281101e-07, 1.0079995263367891e-07, 9.959994349628687e-08, 1.0050006676465273e-07, 1.1399993672966957e-07, 1.4040002133697272e-07, 1.1280004400759936e-07, 1.1869997251778841e-07, 1.121999230235815e-07, 1.078000059351325e-07],
        [5.387000273913145e-07, 5.06900018081069e-07, 5.011000903323293e-07, 4.847999662160873e-07, 4.950999282300472e-07, 5.84700028412044e-07, 5.784000968560576e-07, 6.38199970126152e-07, 6.685000844299794e-07, 7.805000059306622e-07, 1.055200002156198e-06],
        [1.7659994773566723e-07, 1.6439997125416993e-07, 1.6759999562054873e-07, 1.639999682083726e-07, 1.7749995458871126e-07, 1.6599998343735934e-07, 1.727000344544649e-07, 1.7220003064721822e-07, 1.7169991042464972e-07, 1.661001006141305e-07, 1.7159990966320037e-07],
    ]

    for i, test in enumerate(TESTS):
        name, _ = test

        plot = pd.DataFrame({"main": main_results[i], "fixed": fixed_results[i]}, index=SUBMOBJECT_COUNTS).plot.line(title=name)
        plot.set_xlabel("Number of Submobjects")
        plot.set_ylabel("Time per Call")
        plt.savefig(name.strip("_"))


if __name__ == "__main__":
    # uv add https://github.com/ManimCommunity/manim.git
    # uv add https://github.com/nikolajmunk/manim.git --branch perf/faster-mobject-iterable
    # dump_timings()
    plot_timings()
```
</details>

File: manim/mobject/mobject.py (hunk 4 above)
Comment (nikolajmunk) [reply manim/mobject/mobject.py:2516]: After some thinking, I'm inclined to keep the current `chain` implementation. IMO, `chain([self] if self.has_points() else [], self.submobjects)` more clearly communicates what we expect to receive by calling `iter(mob)`: a stream of all the eligible mobjects as per `split`. The higher speed is a small bonus on top.

Happy to hear arguments for/against from everyone else :)

File: manim/mobject/graphing/probability.py (hunk 3 above)
Comment (behackl) [reply manim/mobject/graphing/probability.py:215]: Maybe as a separate project, but I suppose this would need some more discussion. 🤔

File: manim/mobject/mobject.py (hunk 4 above)
Comment (behackl) [reply manim/mobject/mobject.py:2516]: We can keep this as is, its fine. 👍

**Review by behackl — APPROVED** [review]
Excellent! Thanks for the efforts with this, I don't have any further comments. Happy to get this merged!
